src/components/shop-app.js

```diff
@@ -0,0 +1,507 @@
+        break;
+      case 'detail':
+        await import('../components/shop-detail.js');
+        if (this._item) {
+          store.dispatch(updateMeta({
+            title: this._item.title,
+            description: this._item.description.substring(0, 100),
+            image: document.baseURI + this._item.image
+          }));
+        }
+        break;
+      case 'cart':
+        await import('../components/shop-cart.js');
+        store.dispatch(updateMeta({ title: 'Cart' }));
+        break;
+      case 'checkout':
+        await import('../components/shop-checkout.js');
+        store.dispatch(updateMeta({ title: 'Checkout' }));
+        break;
+      default:
+        store.dispatch(updateMeta({ title: '404' }));
+    }
+
+    this._ensureLazyLoaded();
+    if (oldPage) {
+      // The size of the header depends on the page (e.g. on some pages the tabs
+      // do not appear), so reset the header's layout when switching pages.
+      timeOut.run(() => {
+        const header = this.shadowRoot.querySelector('#header');
+        header.resetLayout();
+      }, 1);
+    }
+  }
+
+  _ensureLazyLoaded() {
+    // load lazy resources after render and set `loadComplete` when done.
+    if (!this.loadComplete) {
+      afterNextRender(this, () => {
+        import('./lazy-resources.js').then(() => {
+          // Register service worker if supported.
+          if ('serviceWorker' in navigator) {
+            navigator.serviceWorker.register('service-worker.js', {scope: '/'});
+          }
+          this.loadComplete = true;
+          // The size of the header depends on the page and whether tabs have loaded,
+          // so reset the header's layout after load completes.
+          timeOut.run(() => {
+            const header = this.shadowRoot.querySelector('#header');
+            header.resetLayout();
+          }, 1);
+        });
+      });
+    }
+  }
+
+  _updateLocation() {
+    store.dispatch(updateLocation(window.decodeURIComponent(window.location.pathname)));
+
+    // Close the drawer - in case the *route* change came from a link in the drawer.
+    this.drawerOpened = false;
+  }
+
+  _offlineChanged(offline, oldOffline) {
+    // Show the snackbar if the user is offline when starting a new session
+    // or if the network status changed.
+    if (offline || (!offline && oldOffline === true)) {
+      this._snackbarOpened = true;
+      window.clearTimeout(this._snackbarTimer);
+      this._snackbarTimer = window.setTimeout(() => this._snackbarOpened = false, 4000);
+    }
+  }
+
+  _setMeta(attrName, attrValue, content) {
+    let element = document.head.querySelector(`meta[${attrName}="${attrValue}"]`);
+    if (!element) {
+      element = document.createElement('meta');
+      element.setAttribute(attrName, attrValue);
+      document.head.appendChild(element);
+    }
+    element.setAttribute('content', content || '');
+  }
+  
+  _metaChanged(detail) {
+    // Announce the page's title
+    if (detail.title) {
```

Comment:
This stuff is really rendering so should be done there. Idea is that it's easier to think about that way. It's fine to make a helper that's called in render.